Add an alternative implementation of Delete, Change, Yank operators

badranX opened the pull request:
Fixes #40, Fixes #29, Fixes #110, and Fixes #111.

This is an alternative implementation of the Delete, Change, Yank operators. I kept the previous implementation unchanged, as I think it's more isolated from upstream code (thus it's better but requires a lot of effort to complete it). _However, currently we have only two lines in upstream code, which makes this PR very manageable._

**Example of added commands:**
  - `di)`, `da{`, `dt*`, `dT*`...etc
  -  register control like `"+3yy`
  -  `<Space>y[any combination]` copy to clipboard. Doesn't interfere with the previous Helix `<Space>y` feature.
  
**Example of new commands inherited from Helix (superset of Vim features):** `daf` deletes a function, `yaf` yank a function ...etc

 ### Differences from Vim?
   - No UI visual/cursor change.
   - Sometimes the cursor position after/before applying operators is different (_in non-essential places and not urgent issue_). 
   -  Hitting `v` will stop the operator. Hitting `V` will continue doing visual line mode. These commands are kind of part of evil-helix code, and thus I can add a "check" in them to match the expected behavior. However, I think it's extremely not important to match this behavior.
   
   **Edit:** I'll leave improving `df`, `dt`...etc . Why? Improving `t*`, `f*` movements will directly improve the operator commands of this PR without introducing changes.
   
### Testing
Passes all tests in  https://github.com/thomie/vim-keymap-test/pull/1/files

usagi-flow committed:
doc(evil): add the "VIS" mode to the list of changed defaults

usagi-flow committed:
doc(evil): add repology badges

usagi-flow committed:
doc(evil): inner word motion and some other stuff

usagi-flow committed:
doc(evil): add information regarding the aur package

usagi-flow committed:
doc(evil): fix badge

usagi-flow committed:
doc(evil): add repology badges

usagi-flow committed:
doc(evil): clarify the current state

usagi-flow committed:
doc(evil): fix links

usagi-flow committed:
fix(ci): disable automatic triggering of upstream workflows

usagi-flow committed:
feat(ci): add workflow names

usagi-flow committed:
feat(ci): separate check/test and build workflows

usagi-flow committed:
fix(ci): artifact paths

usagi-flow committed:
fix(ci): explicitly provide a target

usagi-flow committed:
Squashed commit of the following:

commit a7501433a56786fe1b28158f1c414325a20d4e96
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Wed May 15 20:56:11 2024 +0200

    doc(evil): add a notice regarding the branch being rebased

commit 88e3b5f944310d2dd7f7890c58c5dea49d8d2cc9
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Wed May 15 20:56:11 2024 +0200

    doc(evil): add a screenshot

commit 1227eebeb38617a3bd844031ffadb82660037093
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Wed May 15 20:56:11 2024 +0200

    doc(evil): fix links in README.md

commit 5ca6c2ca8fb8a6de2a44ed52be1c1fd533fa2f17
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Wed May 15 20:56:11 2024 +0200

    doc(evil): Update README.md

usagi-flow committed:
Squashed commit of the following:

commit 13654f976ac7be5b11dbb67dfc732974c65fc42d
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sat May 18 00:03:44 2024 +0200

    feat(ci): pr test/build workflow

commit 80f3586f837728d05907acdc54ec8989f7ffe772
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sat May 18 00:03:44 2024 +0200

    fix(ci): x86_64 macos build

commit 26d458b82160bd0ea69ee72df6675e55915465b8
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Thu Apr 25 20:44:51 2024 +0200

    fix(ci): build and create a release when a release tag is pushed

commit b89d1e3ea6ca345c2d80b3d186a8093d4a1d991f
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Thu Apr 25 20:44:51 2024 +0200

    fix(ci): mac builds

commit 4fee940bed8f50d4d43d926751a92e928c099c94
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Thu Apr 25 20:44:51 2024 +0200

    feat(commands): ci

usagi-flow committed:
Squashed commit of the following:

commit f1077392da32385fcb12d888970778b0e6f8b286
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Mon Jul 18 16:43:24 2022 +0200

    feat(tui): minimalistic vertical window separator

usagi-flow committed:
Squashed commit of the following:

commit 3c50f91cbdecb09ec2b9495ba4d8cb2471f3bd96
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Fri Oct 21 22:52:20 2022 +0200

    feat(statusline): if color_modes is enabled, color the file type as well

usagi-flow committed:
Squashed commit of the following:

commit 33162d42417cbce278d469bfaf94753aec1323a7
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Thu May 16 23:28:26 2024 +0200

    feat(defaults): use `catppuccin_macchiato` as the default theme

commit b719768b4f9979400ac54a0c4f390734f136374d
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Thu May 16 23:28:26 2024 +0200

    fix(defaults): smart tab shouldn't be enabled by default

usagi-flow committed:
Squashed commit of the following:

commit b4d065f2aeebb30f3465f88386fcf0591874c015
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Thu Apr 25 20:44:51 2024 +0200

    feat(commands): implement partial line-based selection

commit 62881a8d673e868f3559343c8c5cd94dd9733ee1
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sun Feb 11 15:44:20 2024 +0100

    feat(commands): introduce an `evil` configuration option

    Introduce an `evil` configuration option and respective defaults

commit 7217fadc7cc66409079bc79aef3d3cb371822a1e
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sun Feb 11 15:44:20 2024 +0100

    feat(commands): ensure the last character remains selected when pressing 'b' in select mode

commit 2852cf8b426c8b5480c7876b0f0c2d370682af30
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Fri Mar 31 19:17:20 2023 +0200

    feat(commands): avoid selections in the normal/insert mode

commit c8f4b110cd1652283642b990d7f8197201536e14
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Fri Mar 31 18:21:13 2023 +0200

    feat(command): don't crash when using an unsupported motion key

commit 06dee1c3b35f66234b35739e2b55dde532a30504
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Fri Mar 31 17:58:38 2023 +0200

    feat(commands): avoid crash when using a count after the action key which includes 0

    - The 0-key would be incorrectly interpreted as a motion key.

commit 72e2cd6d823fcbb9683e7d42f3cdfa63e475a85f
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Fri Mar 31 14:35:41 2023 +0200

    feat(commands): better word movement

    - Do not move the anchor when performing a word-based movement in selection mode.

commit 51833cd11d28f966f5f86ca08f078276645af87b
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sun Mar 26 17:01:12 2023 +0200

    feat(commands): implement "delete_immediate" (x)

commit e10ce6e62f82fa19c16b448fcd10ebdc81409de9
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sat Nov 19 15:00:53 2022 +0100

    feat(commands): adjust writing to the register following upstream API changes

commit de0c9c2417b5dadcefbe5597381c5fa5bb971a4c
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sat Oct 22 14:01:11 2022 +0200

    feat(commands): evil-specific movement (no normal-mode selection)

commit 750b68ddea57a100e02a18dae60f6fd4ea48aba6
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sat Oct 22 00:27:05 2022 +0200

    feat(commands): initial change (replace) implementation

commit 9925128826a80b881b40337ac777d1f1dce00d82
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Fri Oct 21 22:42:30 2022 +0200

    feat(commands): word-based selection

commit d37bd99215f25e09da12693365afc21b3b51bffc
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Wed Oct 19 22:00:39 2022 +0200

    feat(commands): initial word-based selection

commit e6bc49e433d6cd43599f694d297b3aa3c23f4119
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sun Oct 16 12:43:44 2022 +0200

    feat(commands): prepare for word-based motions

commit 548333f32762c019bb92f85a43bf74e8781356c1
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sat Oct 1 15:14:14 2022 +0200

    feat(commands): code adjustments following API changes

commit aeb0bf7ecafc4472848e1571bb8fc2c3db91ec77
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Tue Aug 16 23:07:25 2022 +0200

    feat(commands): delete command, refactoring

commit f595eedb0571cd17e04a7e241019f80a8f7ac941
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sun Aug 14 14:39:36 2022 +0200

    feat(commands): better count handling in evil mode

commit b445a1c4fb9d4400cc57d9ac7605dd7be0794788
Author: usagi-flow <2804556+usagi-flow@users.noreply.github.com>
Date:   Sat May 14 18:25:21 2022 +0200

    feat(commands): evil mode

usagi-flow committed:
Squashed commit of the following: (#3)

commit cdc53f42e8e9eeb19f7411bd3d026d75188cd8f5
Author: Jesse Luehrs <doy@tozt.net>
Date:   Sun Sep 17 13:25:52 2023 -0400

    add a separate "helix:" modeline for setting helix-specific config

    for example, if the language name differs between vim and helix

commit 01554579e9ffdb2922b68bc55a42c36be76d3f4f
Author: Jesse Luehrs <doy@tozt.net>
Date:   Sun Sep 17 12:35:57 2023 -0400

    add basic modeline support

    currently only supports setting language, indent style, and line endings

usagi-flow committed:
feat(evil): integrate the latest tag name in the app version info

usagi-flow committed:
feat(evil): adjust the app name in the version info (#6)

usagi-flow committed:
fix(evil): ensure evil mode is enabled by default (#7)

When starting without a user configuration, the editor would load with `evil` set to `false`.
Fix this by not only adjusting the defaults, but by separating the defaults in respective `default_evil()` functions.
The editor is now evil by default.

usagi-flow committed:
feat(evil): implement the "inner word" modifier (#12)

usagi-flow committed:
feat(defaults): statusline layout & colors and make `insert_final_newline` opt-in (#13)

* feat(defaults): statusline layout and colors

Adjust the default layout and enable mode/file-type colors by default

* refactor(defaults): treat the different smart-tab default as an evil-specific default

* feat(defaults): disable `insert_final_newline` by default

This is not a decision that the editor should take for the user. Make it opt-in instead.

usagi-flow committed:
test(defaults): fix `keys_resolve_to_correct_defaults` (#15)

* test(defaults): fix `keys_resolve_to_correct_defaults`

* test(defaults): fix `parsing_keymaps_config_file`

usagi-flow committed:
feat(evil): make find_char commands evil (#14)

usagi-flow committed:
fix(evil): avoid panic when deleting a character (`x`) at the EOF (#23)

usagi-flow committed:
fix(evil): find_char commands shouldn't collapse selections in visual mode (#24)

usagi-flow committed:
feat(evil): implement the word end (`e`) motion (#25)

usagi-flow committed:
feat(evil): implement the long word motions (`W`, `B`, `E`) (#26)

* feat(evil): implement the long word motions (`W`, `B`, `E`)

* feat(evil): add support for word motions `e`, `W`, `B` and `E` when used with a command

usagi-flow committed:
fix(evil): avoid a panic when appending at the EOF (#31)

usagi-flow committed:
feat(evil): bind `match_brackets` to `%` (#32)

usagi-flow committed:
Merge branch 'master'

usagi-flow committed:
chore(evil): clarifications (#41)

* doc(evil): remove the mention of a different default theme

* refactor(evil): improve logging to avoid misconceptions

* doc(evil): we don't rebase that often anymore

* doc(evil): document the `editor.evil` option

usagi-flow committed:
feat(evil): change line (#45)

* feat(commands): implement change line with `cc`

* refactor(commands): logging

usagi-flow committed:
feat(ci): support windows builds (#39)

* fix(treesitter-build): mingw doesn't support `-z`

* feat(ci): support windows builds

* feat(ci): support windows builds

* feat(ci): support windows builds (attempt building with cargo-xwin)

* feat(ci): support windows builds (attempt building with cargo-zigbuild)

* feat(ci): support windows builds (attempt building with cargo-zigbuild)

* feat(ci): support windows builds (attempt building with cargo-zigbuild)

* feat(ci): support windows builds (attempt building with cargo-zigbuild)

* feat(ci): support windows builds (attempt building with cross + clang)

* fix(ci): ensure relro is enabled for linux targets only

See also: https://www.redhat.com/en/blog/hardening-elf-binaries-using-relocation-read-only-relro

* fix(ci): ensure relro is enabled for linux targets only

* fix(ci): ensure relro is enabled for linux targets only

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

* feat(ci): support windows builds (attempt building on the windows runner)

usagi-flow committed:
fix(ci): windows builds require native msvc builds

usagi-flow committed:
fix(ci): windows builds

usagi-flow committed:
fix(ci): separate the release steps as a separate job

usagi-flow committed:
fix(ci): separate the release steps as a separate job

usagi-flow committed:
fix(ci): separate the release steps as a separate job

usagi-flow committed:
Merge tag 25.01

usagi-flow committed:
fix(evil): adjustments following the newly introduced callback kind

thomie committed:
fix(evil): make `w`, `W`, `e`, `E`, `b` and `B` in normal and select mode behave exactly like Vim (#69)

* Simplify evil_move_word_impl, fixing bugs

The function `evil_move_word_impl` can be simplified, thereby fixing the
following two bugs in the process:

1. In select mode, the key sequence `wb` would result in a selection,
   instead of being a no-op as in Vim.

2. The functions `evil_next_word_end` and `evil_next_long_word_end`
   (mapped to 'e' and 'E') in normal mode would make the cursor land on
   the space after the word, instead of the last character of the word.

The only problem with this commit is that now the functions
`evil_next_word_start` and `evil_next_long_word_start` (mapped to `w`
and `W`) don't behave correctly anymore, making the cursor land
on the space before the next word. We'll fix it in the next commit.

* Fix `w` and `W`

1. Space before word: when the cursor is on the space before a word, and
   you hit `w`, Vim will move the cursor to this word, but evil-helix
   would skip the word and move to the _next_ word.

2. Punctuation: when the cursor is on the letter 'f' in "foo.bar qux",
   and you hit `w` twice, Vim will move the cursor to the letter 'b',
   but evil-helix would skip to the letter 'q'.

   This bug has the same underlying cause as the previous one, and the
   fix for both is to move the cursor one character to the left before
   doing the word move.

3. Regression: the previous commit, simplifying `evil_move_word_impl`,
   introduced a bug where `w` would make the cursor land on the space
   before the next word, instead of _on_ the next word. Fix it by moving
   the cursor one character to the right again after the word move.

4. Newlines: when the cursor is on the 'f' in "foo\nbar", and you hit
   `w`, Vim will move the cursor to the letter 'b' (so to the next
   line), but evil-helix would move the cursor onto the newline
   character. See the changes in `reached_target` for the fix.

Fixes #44

thomie committed:
feat(evil): respect editor.default-yank-register (#70)

The configuration option `default-yank-register` was added to Helix in
https://github.com/helix-editor/helix/pull/11430.

Evil helix issue: #53.

usagi-flow committed:
feat: new movements (#66)

* add `same_line` movement commands (#768)

Commands added:
* `move_same_line_char_left`
* `move_same_line_char_right`
* `extend_same_line_char_left`
* `extend_same_line_char_right`
* `append_mode_same_line`

These new commands move cursors, while making them stay in the same
line. So if a cursor would wrap around into another line, instead it
won't move and stay at its current position.

* add `anchored` movement commands

Commands added:
* `move_anchored_line_up`
* `move_anchored_line_down`
* `move_anchored_visual_line_up`
* `move_anchored_visual_line_down`
* `extend_anchored_line_up`
* `extend_anchored_line_down`
* `extend_anchored_visual_line_up`
* `extend_anchored_visual_line_down`

These new commands move cursors vertically. A cursor will move depending
on its position:

* If it is on a newline character of a non-empty line, the cursor will
  stay on newlines (i.e. on a line's last character).

* If it is on a non-newline character of a non-empty line, the cursor
  will try to avoid newline characters. It will move normally, but if
  it would end up on a newline, instead it will be moved one position
  left of it (i.e. the line's second to last character).

* If it is on the newline character of an empty line (that contains
  nothing except the newline character), the cursor will continue to
  move like before: If it stayed on newline before, it will continue to
  do so. Otherwise it will try to avoid them (except on empty lines).

* feat(evil): make the new (vim-like) movement default

* feat(evil): ensure selections are collapsed when appending

---------

Co-authored-by: Pantos <pantos+github@akk.org>

usagi-flow committed:
feat: colored/rainbow indentation guides (#76)

* Squashed commit of the following:

commit b5bf68129b56dc0178a61b7e22151decf82a25da
Merge: d1517c6a 207829ee
Author: SoraTenshi <dream@neoncity.dev>
Date:   Thu May 25 19:53:22 2023 +0200

    Merge branch 'master' into colored-indent-guides

commit d1517c6a4b9b09609df03d984f4e890ec1998c26
Author: SoraTenshi <dream@neoncity.dev>
Date:   Sun Feb 12 22:18:41 2023 +0100

    Update to new TextRenderer

    Docs

    fmt

commit 7f47756b6d3d9119e1e426d7df48221d9019265c
Author: Michael Davis <mcarsondavis@gmail.com>
Date:   Mon Jul 11 19:10:57 2022 -0500

    Parse rainbow style array in themes

    This change adds a field to the schema of themes which takes a
    list of styles.

        rainbow = ["red", "orange", "yellow", { modifiers = ["reversed"] }]
        [palette]
        red = "#ff0000"
        orange = "#ffa500"
        yellow = "#fff000"

    Normal style rules apply for each element in `rainbows`: you can
    use definitions from the palette and the full fg/bg/modifiers
    notation.

    Themes written with `rainbow` keys are not backwards compatible.
    Parsing errors will be generated for older versions of Helix
    attempting to use themes with `rainbow` keys.

    A default rainbow is provided with base16 colors.

    This change is made with rainbow pair characters (parens, brackets, etc.)
    in mind but it could also be used for other rainbow cosmetic elements
    like rainbow indent-guides.

* fix: avoid panic, properly set rainbow_length when loading theme

* feat: define a rainbow palette for the catppuccin themes

usagi-flow committed:
doc(evil): readme

badranX committed:
feat(evil): add vi/va keymaps (#78)

thomie committed:
feat(evil): visual line mode (#82)

* Add scaffolding for visual line mode

Doesn't do anything yet.

* Implement visual line mode

badranX committed:
feat(evil): add caret (^) keymap (#84)

badranX committed:
feat(evil): add `}` and `{` keymaps (#83)

* feat(evil): map } and { to ]p and [p without selection

* feat(evil): add } and { vim movements

antoyo committed:
feat(evil): allow G to take a count, jumping to the line specified in the count (#93)

badranX committed:
feat(evil): add delete, change, and yank operators

This is an altrnative implementation to the one we have. It uses basic
Helix select mode to simulate Vim operators.

badranX committed:
refactor: better naming & func arguments

badranX committed:
perf: avoid EvilOps tests unless pending an operator

badranX committed:
fix: vi<Esc> doesn't collapse selection in Vim/Helix

antoyo committed:
feat(evil): add support for indent and unindent in insert mode (#97)

badranX committed:
refactor: undo changes in some upstream commands

Since post_command_execution string matching only happens when pending
an operator, performance issues are minimal.

antoyo commented:
Hey.
I saw you also implemented the feature I implemented in #101.

One difference I see between our implementations is that mine will show the shortcuts like Helix usually do:
![screenshot](https://github.com/user-attachments/assets/f0b34f89-2d07-4d4b-b93e-e506ae4e751d)

Do you think it would be easy to have this behavior in your PR?
Thanks.

badranX commented:
> Hey.
> I saw you also implemented the feature I implemented in #101.
> 

Sorry, didn't notice. I actually didn't implement it, it's just a side effect of this PR. I think it was originally implemented by @usagi-flow . 

> Do you think it would be easy to have this behavior in your PR?

Yes, it's possible, but I don't want to introduce more changes yet beyond the core features as I'm not sure if @usagi-flow would accept the core idea of this PR.

badranX committed:
fix: stop pending operator if not Select mode

badranX committed:
refactor: operator_impl take register argument

badranX committed:
feat: add yank_to_clipboard evil command

This is not a Vim feature, but we find it in many Vim extensions. It
acts as a superset, doesn't interfere with what's expected from
Vim/Helix.

badranX committed:
refactor: move funcs to evil.rs

badranX committed:
refactor: reduce changes to upstream code

Current changes:
  - one line in `Commands::select_textobject`
  - one line in `Commands::MappableCommand::execute`
  - defining short evil_* commands.

badranX committed:
fix: force select-mode evil yank to use operator register

badranX committed:
refactor: allow some commands to take custom register

- Allow some EvilCommands to take custom register argument
- A lot of changes to simplify code

badranX committed:
fix: stay on same line when cc empty line (old bug)

badranX committed:
feat: add comands like dj, cj, dw

badranX committed:
fix: operators work in Visual Line selection mode

badranX committed:
feat: proper support for [d|y|c]{N}[G|gg]

badranX committed:
feat(evil): search word under cursor with `*` / `#` (#79)

* feat(evil): add Vim-style * and # key search

* fix(evil): match */# search by word boundaries

Co-authored-by: Thomas Miedema <157521+thomie@users.noreply.github.com>

* feat(evil): stop */# search if selection is empty

If cursor is over an empty string, */# will do nothing except displaying
a message, "No string under cursor".

* fix(evil): */# is case insensitive if smart-case is true

If search.smart-case is true then search is case insensitive (like Vim)
otherwise it's case senstive

* feat(evil): */# searches the nearest non-blank word

If cursor on blank, */# searches for the next closest word in the line.

* feat(evil): */# searches the nearest 'keyword'

Here, a 'keyword' is a helix word that starts with a char, ch, that satisfies:
	ch.is_alphanumeric() || ch == '_' charachter

feat: */# searches the nearest 'keyword'

* fix(evil): */# ignore helix very last line

* refactor(evil): remove redundant movement logic in */#

Clarified that Helix 'word' is used as the equivalent of Vim's default
'keyword'. Also removed the unused Movement parameter.

* fix(evil): */# searches only primary cursor/selection

Multi cursor can cause different unwanted behavior, including panicking
on Helix very last line (~).

* perf(evil): */# test a single char under cursor

* feat(evil): */# exit select mode like Vim

* feat(evil): make */# search always case-insensitive

Achieved by prepending The regex command (?i) to search

* refactor(evil): format code

* refactor(evil): comments

---------

Co-authored-by: Thomas Miedema <157521+thomie@users.noreply.github.com>
Co-authored-by: 兎の流れ <2804556+usagi-flow@users.noreply.github.com>

usagi-flow commented on helix-term/src/commands/evil.rs:
```diff
@@ -3,28 +3,41 @@ use std::{
+pub enum EvilOpsCase {
+    CompleteLines,
+    NextWord,
+}
```
Could you please add a little bit of `/// documentation` to explain what this enum covers?

usagi-flow commented on helix-term/src/commands/evil.rs:
```diff
@@ -862,3 +940,213 @@ pub fn evil_movement_paragraph_forward(
     Range::new(anchor, head)
 }
+
+pub struct EvilOps;
```
Could you please add a little bit of `/// documentation` to briefly describe this (very important) struct?

usagi-flow commented on helix-term/src/commands/evil.rs:
```diff
@@ -862,3 +940,213 @@ pub fn evil_movement_paragraph_forward(
+            return;
+        }
+
+        // The following will only run if operator is pending
+        match cmd.name() {
+            "select_textobject_around"
+            | "select_textobject_inner"
```
It would be amazing if we could avoid a string comparison.
I have no good idea that immediately comes to my mind; mappable commands do not have an ID, Hash or anything else that can be compared efficiently. Changing the structure of `MappableCommand` is quite invasive. Wrapping the command in something like an `EvilMappableCommand` (so that we can associate more info to commands) could work, but could cause more work elsewhere...

antoyo replied:
A solution to this would be to do like I did in [my own PR](https://github.com/usagi-flow/evil-helix/pull/101) (which took inspiration from the [other evil helix fork](https://github.com/RoastBeefer00/evilhelix)) that implements one command of the current PR.
With this, I believe there will be no need for the `EvilOps` struct.

This would have the additional benefit of [showing the keybindings help window](https://github.com/usagi-flow/evil-helix/pull/103#issuecomment-2898045325) as is usual in helix.

usagi-flow reviewed (commented):
This is a very interesting PR; alone due to the fact that it could refactor the initial implementation of evil-helix, which I hacked together at a time where I didn't think this project would ever be this popular.

I need some more time to carefully look at the changes and their impact, but I think this is something we can work with.

I hope I'll find the time soon to provide some initial tests; I think that's important if we're digging into the core of evil-helix.

badranX committed:
Merge branch 'usagi-flow:main' into dev-evil-operators

badranX committed:
docs: add documentation for EvilOps and EvilOpsCase

hujianxin commented:
@usagi-flow evil-helix saves my coding life, Thanks for the great work. 

And any progress for this PR, do you still follow this?
